[AArch64][Machine-Combiner] Split loads into lanes of neon vectors into multiple vectors when possible

jcohen-apple opened the pull request:
This changes optimizes gather-like sequences, where we load values separately into lanes of a neon vector. Since each load has serial dependency, when performing multiple i32 loads into a 128 bit vector, it is more profitable on wide processors to load into separate vector registers and zip them. This shows a small speedup on internal workloads at Apple.

jcohen-apple committed:
Initial unit test to demonstrate current behavior

jcohen-apple committed:
Apply pattern to basic case of 4 i64 loads into fpr128 register

davemgreen commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -7317,6 +7319,57 @@ static bool getMiscPatterns(MachineInstr &Root,
+      Root.getMF()->getSubtarget().getRegisterInfo();
+  // Enable this only on Darwin targets, where it should be profitable. Other
+  // targets can remove this check if it is profitable there as well.
+  if (!Root.getMF()->getTarget().getTargetTriple().isOSDarwin())
```
CPU tuning subtarget features are the preferred way to tune for different systems. MachineCombiner can use scheduling depths to calculate when it should be profitable, so you might be able to enable it more generally. It feels like it might be OK as a DAG combine to be honest, although it would add more instructions IIUC.

jcohen-apple replied:
Ack, removed it.
I added this as a machine combiner pattern because I want to use the checks on resource length and reducing critical path before applying, since it does add additional instructions.

davemgreen commented on llvm/test/CodeGen/AArch64/aarch64-combine-split-loads.mir:
```diff
@@ -0,0 +1,52 @@
+    ; CHECK-NEXT: [[LD_i32:%[0-9]+]]:fpr32 = LDRSroX [[COPY]], killed [[COPY1]], 0, 1
+    ; CHECK-NEXT: [[FIRST_REG:%[0-9]+]]:fpr128 = SUBREG_TO_REG 0, killed [[LD_i32]]
+    ; CHECK-NEXT: [[SECOND_REG:%[0-9]+]]:fpr128 = IMPLICIT_DEF
+    ; CHECK-NEXT: [[LD1_0:%[0-9]+]]:fpr128 = LD1i32 [[SECOND_REG]], 0, killed [[COPY2]] 
```
Using LDRSroX could help break the dependency on whatever register register %SECOND_REG was randomly chosen.

jcohen-apple committed:
Support additional data types

jcohen-apple committed:
Remove check for isOsDarwin()

jcohen-apple committed:
Formatting changes

davemgreen commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -7367,11 +7372,241 @@ static bool getMiscPatterns(MachineInstr &Root,
+  auto *CurrInstr = MRI.getUniqueVRegDef(Root.getOperand(1).getReg());
+  auto Range = llvm::seq<unsigned>(1, NumLanes - 1);
+  SmallSet<unsigned, 4> RemainingLanes(Range.begin(), Range.end());
+  while (RemainingLanes.begin() != RemainingLanes.end() &&
```
RemainingLanes.empty().
Check CurrInstr is not null, just in case.

jcohen-apple replied:
Thanks

davemgreen commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -7327,6 +7329,9 @@ bool AArch64InstrInfo::isThroughputPattern(unsigned Pattern) const {
   case AArch64MachineCombinerPattern::MULSUBv2i32_indexed_OP2:
   case AArch64MachineCombinerPattern::MULSUBv4i32_indexed_OP1:
   case AArch64MachineCombinerPattern::MULSUBv4i32_indexed_OP2:
+  case AArch64MachineCombinerPattern::GATHER_i32:
```
Can we call them LANEWISEGATHER or GATHERLANE or something like that, to distinguish from the gather instructions that we have.

jcohen-apple replied:
Done

davemgreen reviewed (commented):
Looks decent.

Should it bail out for minsize?
Should it support v8i8 / v41i6 too?

jcohen-apple committed:
Code review comments

- Early exit if optimizing for size
- Fix loop condition to check if CurrInstr is not null
- use .empty() instead of begin() != end()
- Rename pattern enum

jcohen-apple commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -7367,11 +7372,241 @@ static bool getMiscPatterns(MachineInstr &Root,
+}
+
+/// Search for patterns where we use LD1 instructions to load into
+/// separate lanes of an 128 bit Neon register. We can increase MLP
```
I would think My Little Pony would be agnostic to this optimization, but I haven't checked

There is the ILP aspect, but IIUC the LD1 instructions introduce a data dependency that prevents the loads to lane 1-3 from being issued until the previous lane completes. So by splitting the load you gain ILP but can also issue the loads (slightly) sooner, and with more parallelism.

Removed the acronym from the comment just to be safe :)

jcohen-apple commented:
Added an early exit for minsize, good catch.
Regarding supporting other ZIP types - technically you can, I couldn't find a case where it was actually profitable. I could address in a follow-up PR if you'd like.

jcohen-apple committed:
fix unit-tests

jcohen-apple committed:
regenerate checks for affected llc tests

fhahn commented on llvm/test/CodeGen/AArch64/aarch64-combine-split-loads.mir:
```diff
@@ -0,0 +1,268 @@
+    %3:gpr64common = COPY $x3
+    %4:gpr64common = COPY $x4
+    %5:fpr32 = LDRSroX %0, killed %1, 0, 1
+    %6:fpr128 = SUBREG_TO_REG 0, killed %5, %subreg.ssub
```
It looks like we only have positive tests here. We should probably have tests where we 
* do not load all lanes, 
*  load all lanes but the order of the lanes is mixed, 
* the first instruction for lane 0 isn't a SUBREG_TO_REG
* some of the LD1 have multiple users

jcohen-apple replied:
There was a negative test at the bottom of the file (for not loading all lanes)
Added the other tests you suggested.
For the out of order lanes - I added that as a positive test case, and fixed the pattern.
Added an extra negative test case where the load to lanes have complex addressing mode (which I don't cover)

fhahn commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -7367,11 +7372,245 @@ static bool getMiscPatterns(MachineInstr &Root,
+                             SmallVectorImpl<unsigned> &Patterns,
+                             unsigned LoadLaneOpCode, unsigned NumLanes) {
+  // Early exit if optimizing for size.
+  if (Root.getMF()->getFunction().hasMinSize())
```
Might be worth pulling out Root.getMF() into a separate variable

fhahn commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -7367,11 +7372,245 @@ static bool getMiscPatterns(MachineInstr &Root,
+
+  // Check that we have load into all lanes except lane 0.
+  // For each load we also want to check that:
+  // 1. It has a single debug use (since we will be replacing the virtual
```
```suggestion
  // 1. It has a single non-debug use (since we will be replacing the virtual
```

fhahn commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -7367,11 +7372,245 @@ static bool getMiscPatterns(MachineInstr &Root,
+      Opcode = AArch64::LDRHui;
+      break;
+    case 16:
+      Opcode = AArch64::LDRBui;
```
For 8 and 16 lanes, it may be worth splitting into more than 2 separate chains of loads?

jcohen-apple replied:
Like I wrote above to @davemgreen - it kicks in less often because of the additional use of load ports. I wanted to keep the pattern (relatively) simple, so decided not to do that right now. I can add it in a follow up PR if you think it's a good idea.

fhahn replied:
Yeah sounds good as follow-up

jcohen-apple committed:
code review comment - florian

jcohen-apple committed:
rename test file

fhahn reviewed (approved):
LGTM, thanks

jcohen-apple committed:
Merge branch 'main' into users/joncoh-apple/split_neon_loads
